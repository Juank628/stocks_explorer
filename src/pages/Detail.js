import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import styles from './Detail.module.scss';
import { readPerformances } from '../redux/slices/details';

export default function Detail() {
  const [sectorData, setSectorData] = useState({
    name: 'undefined',
    marketCap: 0,
    numberOfCompanies: 0,
  });
  const { sectors } = useSelector((store) => store.stocksReducer);
  const { performances } = useSelector((store) => store.detailsReducer);
  const { sector } = useParams();
  const dispatch = useDispatch();

  const getData = (sectorName, sectorsList) => {
    const formatedSector = sectorName.replace(/-/g, ' ');
    const selectedSector = sectorsList.find((item) => item.name === formatedSector);
    setSectorData(selectedSector);
  };

  const getPerformance = (sectorName, performancesObject) => {
    let formatedSector = sectorName.replace(/-/g, '');
    formatedSector = `${formatedSector[0]
      .toLowerCase() + formatedSector
      .substring(1)}ChangesPercentage`;
    if (performancesObject) return performance[formatedSector];
    return 0;
  };

  useEffect(() => {
    if (performances.length === 0) {
      dispatch(readPerformances());
    }
    getData(sector, sectors);
    getPerformance(sector, performances[0]);
  }, []);

  return (
    <div className={styles.mainContainer}>
      <h3>{sectorData?.name}</h3>
      <div className={styles.metricsContainer}>
        <div className={styles.box}>
          <p>Market cap.</p>
          <p>{`${sectorData?.marketCap}M`}</p>
        </div>
        <div className={styles.box}>
          <p>Companies</p>
          <p>{sectorData?.numberOfCompanies}</p>
        </div>
      </div>
      <p>Top 5 companies</p>
      <ul className={styles.companiesContainer}>
        { sectorData?.topCompanies?.map((company) => (
          <li key={company.symbol} className={styles.box}>
            <p>{company.companyName}</p>
            <p>{company.symbol}</p>
            <p>{`${Math.round(company.marketCap / 1000000, 0)}M`}</p>
          </li>
        )) }
      </ul>
      <p>Industries:</p>
      <ul className={styles.industriesContainer}>
        { sectorData?.industries?.map((sector) => (
          <li className={styles.box} key={sector}>{sector}</li>
        )) }
      </ul>
    </div>
  );
}
