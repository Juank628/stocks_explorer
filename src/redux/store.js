import { configureStore } from '@reduxjs/toolkit';
import stocksReducer from './slices/stocks';
import detailsReducer from './slices/details';

const store = configureStore(
  {
    reducer: { stocksReducer, detailsReducer },
  },
);

export default store;
