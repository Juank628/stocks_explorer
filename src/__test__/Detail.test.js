import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/store';
import '@testing-library/jest-dom';
import Detail from '../pages/Detail';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    sector: 'Technology',
  }),
}));

describe('Home page', () => {
  test('renders home page', () => {
    const element = render(
      <Provider store={store}>
        <BrowserRouter>
          <Detail />
        </BrowserRouter>
      </Provider>,
    );
    expect(element.toJSON).toMatchSnapshot();
  });
});
