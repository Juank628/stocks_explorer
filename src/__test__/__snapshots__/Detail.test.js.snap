// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders home page 1`] = `undefined`;
